### PR TITLE
[REVIEW] Adding limiting adaptor

### LINKED ISSUE #442: [FEA] Memory resource to limit the total amount of memory that can be allocated
**Is your feature request related to a problem? Please describe.**
Spark can run Pandas UDFs, and we would like to support Pandas UDFs that use cudf instead of Pandas.  That means we need the ability to share GPU resources between two processes, JVM and Python.  Eac process normally assumes it owns the entire GPU and potentially can allocate all GPU memory, preventing the other process from allocating sufficient GPU memory to succeed.

One way to support this is by artificially partitioning the GPU memory between the two processes.  If each process only allocates a configured amount then they could successfully share the same GPU.  For processes using RMM, it would be nice if RMM had a device memory resource that can be configured with a total amount of GPU memory it will allow to be allocated.

**Describe the solution you'd like**
A new device memory resource class that takes a constructor parameters of the parent memory resource to use for memory and a total number of bytes as a limit.  The memory resource would track the number of bytes allocated and return allocation failure for any allocation that would exceed the configured total number of bytes that can be allocated.

**Describe alternatives you've considered**
This also could be done with a total pool size config on the pool allocators, but this seems like a more general solution that isn't limited to a specific type of memory resource.  This resource could be stacked with other resources to add a total allocation limit to any type of memory resource.

### PR DESCRIPTION
**Pull request by hyperbolic2346**
This PR creates a limiting adaptor, which allows limiting the amount of memory available. Fixes #442.

Also fixing a small copy/paste error in `logging_resource_adaptor` as well.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by hyperbolic2346** [comment]
Reviews began inside #575. Splitting the limiting adaptor out into its own PR.

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,7 @@
 - PR #560 Remove deprecated `get/set_default_resource` APIs
 - PR #543 Add an arena-based memory resource
 - PR #580 Install CMake config with RMM
+- PR #594 Adding limiting adaptor
```
Comment (harrism) [inline CHANGELOG.md:9]:
```suggestion
- PR #594 Adding limiting memory resource adaptor
```

File: include/rmm/mr/device/limiting_resource_adaptor.hpp [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+  /**
+   * @brief Query the number of bytes that have been allocated. Note that
+   * this can not be used to know how large of an allocation is possible due
+   * to both possible fragmentation and also internal page sizes and alignment
+   * that is not tracked by this allocator.
+   *
+   * @return std::size_t number of bytes that have been allocated through this
+   * allocator.
+   */
+  std::size_t allocated_bytes() const { return allocated_bytes_; }
```
Comment (harrism) [inline include/rmm/mr/device/limiting_resource_adaptor.hpp:98]: I'm undecided on this. Should we prefix getters  with `get_`? `get_upstream` is, and `get_mem_info`. So maybe for consistency.

```suggestion
  std::size_t get_allocated_bytes() const { return allocated_bytes_; }
```

File: include/rmm/mr/device/limiting_resource_adaptor.hpp [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+  /**
+   * @brief Query the maximum number of bytes that this allocator is allowed
+   * to allocate. This is the limit on the allocator and not a representation of
+   * the underlying device. The device may not be able to support this limit.
+   *
+   * @return std::size_t max number of bytes allowed for this allocator
+   */
+  std::size_t allocation_limit() const { return allocation_limit_; }
```
Comment (harrism) [inline include/rmm/mr/device/limiting_resource_adaptor.hpp:107]: ?

```suggestion
  std::size_t get_allocation_limit() const { return allocation_limit_; }
```

**Review by harrism — APPROVED** [review]
Approve, just one question about naming  getters, and a request to add a bit more detail to changelog.

**Comment by hyperbolic2346** [comment]
I like the suggested changes, thank you @harrism.

File: include/rmm/mr/device/limiting_resource_adaptor.hpp [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override
+  {
+    void* p = nullptr;
+
+    std::size_t proposed_size = rmm::detail::align_up(bytes, allocation_alignment_);
+    if (proposed_size + allocated_bytes_ <= allocation_limit_) {
+      p = upstream_->allocate(bytes, stream);
+      allocated_bytes_ += proposed_size;
+    } else {
+      throw rmm::bad_alloc{std::string{"CUDA error: Unable to allocate memory due to limit"}};
```
Comment (jrhemstad) [inline include/rmm/mr/device/limiting_resource_adaptor.hpp:132]:
```suggestion
      throw rmm::bad_alloc{"Exceeded memory limit"};
```

File: include/rmm/mr/device/limiting_resource_adaptor.hpp [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+#include <rmm/detail/error.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+
+namespace rmm {
+namespace mr {
+/**
+ * @brief Resource that uses `Upstream` to allocate memory and limits the total
+ * allocations possible.
+ *
```
Comment (jrhemstad) [inline include/rmm/mr/device/limiting_resource_adaptor.hpp:26]: Doc thread safety.

File: include/rmm/mr/device/limiting_resource_adaptor.hpp [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+   * @brief Construct a new limiting resource adaptor using `upstream` to satisfy
+   * allocation requests and limiting the total allocation amount possible.
+   *
+   * @throws `rmm::logic_error` if `upstream == nullptr`
+   *
+   * @param upstream The resource used for allocating/deallocating device memory
+   * @param allocation_limit Maximum memory allowed for this allocator.
+   */
+  limiting_resource_adaptor(Upstream* upstream,
+                            std::size_t allocation_limit = std::numeric_limits<std::size_t>::max(),
```
Comment (jrhemstad) [inline include/rmm/mr/device/limiting_resource_adaptor.hpp:47]: It feels odd to default the `allocation_limit`. I can't imagine a case where someone is bothering to use `limiting_resource_adaptor` and doesn't want to explicitly set the `allocation_limit`.

File: include/rmm/mr/device/limiting_resource_adaptor.hpp [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+   *
+   * @throws `rmm::cuda_error` if unable to retrieve memory info.
+   *
+   * @param stream Stream on which to get the mem info.
+   * @return std::pair contaiing free_size and total_size of memory
+   */
+  std::pair<size_t, size_t> do_get_mem_info(cudaStream_t stream) const override
+  {
+    auto ret = upstream_->get_mem_info(stream);
+    return {std::min(ret.first, allocation_limit_ - allocated_bytes_),
```
Comment (jrhemstad) [inline include/rmm/mr/device/limiting_resource_adaptor.hpp:188]: This logic doesn't look right. Either it should return what the upstream returns, or only return `allocation_limit - allocated_bytes` and `allocation_limit`. Probably the latter.

**Review by jrhemstad — CHANGES_REQUESTED** [review]

File: include/rmm/mr/device/limiting_resource_adaptor.hpp (hunk 7 above)
Comment (hyperbolic2346) [reply include/rmm/mr/device/limiting_resource_adaptor.hpp:188]: The thought was if you are limited to 8 gigs, but only have a card with 4 gigs that returning 8 gigs was not correct. I'm happy either way though.

Comment (harrism) [reply include/rmm/mr/device/limiting_resource_adaptor.hpp]: Good point. But what if `upstream` does not support `get_mem_info`?

**Review by jrhemstad — APPROVED** [review]

**Pull request merged**